src/front/spv/flow.rs

```diff
@@ -167,36 +191,103 @@ impl FlowGraph {
 
     /// Traverses the flow graph and returns a list of Naga's statements.
     pub fn to_naga(&self) -> Result<crate::Block, Error> {
-        self.naga_traverse(node_index(0))
+        self.naga_traverse(node_index(0), None)
     }
 
-    fn naga_traverse(&self, node_index: BlockNodeIndex) -> Result<crate::Block, Error> {
+    fn naga_traverse(
+        &self,
+        node_index: BlockNodeIndex,
+        stop_node_index: Option<BlockNodeIndex>,
+    ) -> Result<crate::Block, Error> {
+        if let Some(stop_node_index) = stop_node_index {
+            if stop_node_index == node_index {
+                return Ok(vec![]);
+            }
+        }
+
         let node = &self.flow[node_index];
 
         match node.ty {
             Some(ControlFlowNodeType::Header) => {
+                let merge_node_index = self.block_to_node[&node.merge.unwrap().merge_block_id];
+                let mut result = node.block.clone();
+
                 match node.terminator {
                     Terminator::BranchConditional {
                         condition,
                         true_id,
                         false_id,
                     } => {
-                        let mut result = node.block.clone();
                         result.push(crate::Statement::If {
                             condition,
-                            accept: self.naga_traverse(self.block_to_node[&true_id])?,
-                            reject: self.naga_traverse(self.block_to_node[&false_id])?,
+                            accept: self.naga_traverse(
+                                self.block_to_node[&true_id],
+                                Some(merge_node_index),
+                            )?,
+                            reject: self.naga_traverse(
+                                self.block_to_node[&false_id],
+                                Some(merge_node_index),
+                            )?,
+                        });
+                    }
+                    Terminator::Switch {
+                        selector,
+                        default,
+                        ref targets,
+                    } => {
+                        let mut cases = FastHashMap::default();
+
+                        for i in 0..targets.len() {
+                            let left_target_node_index = self.block_to_node[&targets[i].1];
+
+                            let fallthrough: Option<crate::FallThrough> = if i < targets.len() - 1 {
```

Comment:
This could be neatly written as `let fallthrough = targets.get(i+1).map(|target| ...);`